pallets/amm-support/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+		self.0.pop().ok_or(())
+	}
+
+	fn get(self) -> Vec<ExecutionType<IncrementalIdType>> {
+		self.0.into_inner()
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn incremental_id)]
+	pub(super) type IncrementalId<T: Config> = StorageValue<_, IncrementalIdType, ValueQuery>;
+
+	// TODO:
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn id_stack)]
+	pub(super) type IdStack<T: Config> = StorageValue<_, ExecutionIdStack, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		MaxStackSizeReached,
+		EmptyStack,
+	}
+
+	// on initialize - clear operation_id stack if not empty
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade executed.
+		Swapped {
+			swapper: T::AccountId,
+			filler: T::AccountId,
+			filler_type: Filler<AssetId, OtcOrderId>,
+			operation: TradeOperation,
+			inputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+			outputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+			fees: Vec<Fee<AssetId, Balance, T::AccountId>>,
+			operation_id: Vec<ExecutionType<IncrementalIdType>>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Returns next incremental ID and updates the storage.
+	pub fn next_incremental_id() -> Result<IncrementalIdType, DispatchError> {
+		IncrementalId::<T>::try_mutate(|current_id| -> Result<IncrementalIdType, DispatchError> {
+			let inc_id = *current_id;
+			*current_id = current_id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
+			Ok(inc_id)
+		})
+	}
+
+	#[allow(clippy::too_many_arguments)]
+	pub fn deposit_trade_event(
+		swapper: T::AccountId,
+		filler: T::AccountId,
+		filler_type: Filler<AssetId, OtcOrderId>,
+		operation: TradeOperation,
+		inputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+		outputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+		fees: Vec<Fee<AssetId, Balance, T::AccountId>>,
+	) {
+		Self::deposit_event(Event::<T>::Swapped {
+			swapper,
+			filler,
+			filler_type,
+			operation,
+			inputs,
+			outputs,
+			fees,
+			operation_id: <Self as ExecutionTypeStack<IncrementalIdType>>::get(),
+		});
+	}
+}
+
+impl<T: Config> ExecutionTypeStack<IncrementalIdType> for Pallet<T> {
+	fn push(execution_type: ExecutionType<IncrementalIdType>) -> DispatchResult {
+		IdStack::<T>::try_mutate(|stack| -> DispatchResult {
+			stack
+				.push(execution_type)
+				.map_err(|_| Error::<T>::MaxStackSizeReached.into())
```

Comment:
Shouldn't we cap this somehow?

Reply:
We have it capped to 10, by using this const: https://github.com/galacticcouncil/hydration-node/blob/39a2d9a478368b7dd90146b7d387789ac6a5424a/pallets/amm-support/src/lib.rs#L43

Reply:
note : it can be done a much easier way ( let's discuss on call )